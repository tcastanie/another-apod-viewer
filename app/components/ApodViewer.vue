<script lang="ts" setup>
import type { Apod } from '~/types'

const { formattedFullDate } = useDateWidget()

const { date } = defineProps<{ date?: string }>()

const { data, error, status } = await useLazyFetch<Apod>(`/api/apod`, {
  query: { date },
  server: false,
  timeout: 10000,
})

const openImage = async (url?: string) => {
  if (url) {
    await navigateTo(url, { external: true })
  }
}
</script>

<template>
  <div v-if="['pending', 'idle', 'error'].includes(status)">
    <USkeleton class="w-full h-[350px] xl:h-[650px] mb-8" />
    <UAlert
      v-if="error"
      :title="`Error ${error.statusCode} ${error.statusMessage || error.message.includes('aborted') ? 'Nasa API is most likely down' : error.message}`"
      :description="JSON.stringify(error.data, null, 4)"
      icon="i-ph-warning-octagon"
      color="red"
      variant="subtle"
      :ui="{ description: 'whitespace-pre font-mono' }"
    />
    <UProgress
      v-if="['pending', 'idle'].includes(status)"
      animation="swing"
      size="xl"
      :max="['Fetching data from NASA API...']"
    />
  </div>
  <div v-else class="flex flex-col gap-8">
    <template v-if="data?.media_type === 'image'">
      <NuxtImg
        :src="data.url"
        :alt="data.title"
        :class="[{ 'cursor-pointer': data?.hdurl }, 'mx-auto rounded-md']"
        @click="openImage(data.hdurl)"
      />
    </template>
    <iframe
      v-else-if="data?.media_type === 'video'"
      :src="data.url"
      class="w-full rounded-md h-[350px] xl:h-[650px]"
    />
    <div v-if="data" class="flex flex-col">
      <div class="mb-3 text-sm/6 font-semibold text-primary flex items-center gap-1.5">
        {{ formattedFullDate }}
      </div>
      <UiH1>{{ data.title }}</UiH1>
      <UiP v-if="data.explanation" class="mt-4">
        {{ data.explanation }}
      </UiP>
      <UiP v-if="data.copyright" class="mt-4 flex items-center ring-1 ring-gray-800 rounded w-fit px-2">
        <UIcon name="i-ph-copyright" class="inline-block mr-2" />
        <span>{{ data.copyright }}</span>
      </UiP>
    </div>
    <DevOnly>
      <pre class="max-w-full overflow-x-auto">{{ JSON.stringify(data, null, 2) }}</pre>
      <pre class="max-w-full overflow-x-auto">{{ JSON.stringify(error, null, 2) }}</pre>
    </DevOnly>
  </div>
</template>
