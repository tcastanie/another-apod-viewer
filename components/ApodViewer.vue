<script lang="ts" setup>
import type { Apod } from '~/types'

const { formattedFullDate } = useDateWidget()

const { date } = defineProps<{ date?: string }>()

const { data, error } = await useFetch<Apod>(`/api/apod`, {
  query: { date },
})

const openImage = async (url?: string) => {
  if (url) {
    await navigateTo(url, { external: true })
  }
}
</script>

<template>
  <div class="flex flex-col gap-8">
    <DevOnly v-if="0">
      <pre class="max-w-full overflow-x-hidden">{{ JSON.stringify(data, null, 2) }}</pre>
    </DevOnly>
    <UAlert
      v-if="error?.data"
      :title="'Error ' + error.data.code"
      :description="error.data.msg"
      icon="i-ph-warning-octagon"
      color="red"
      variant="outline"
    />
    <template v-if="data?.media_type === 'image'">
      <NuxtImg
        :src="data.url"
        :alt="data.title"
        class="max-h-[650px]"
        :class="[{ 'cursor-pointer': data?.hdurl }, 'mx-auto rounded-md']"
        @click="openImage(data.hdurl)"
      />
    </template>
    <iframe
      v-else-if="data?.media_type === 'video'"
      :src="data.url"
      class="w-full rounded-md h-[650px]"
    />
    <div class="flex flex-col">
      <div class="mb-3 text-sm/6 font-semibold text-primary flex items-center gap-1.5">
        {{ formattedFullDate }}
      </div>
      <UiH1>{{ data?.title }}</UiH1>
      <UiP v-if="data?.explanation" class="mt-4">
        {{ data.explanation }}
      </UiP>
      <UiP v-if="data?.copyright" class="mt-4 flex items-center ring-1 ring-gray-800 rounded w-fit px-2">
        <UIcon name="i-ph-copyright" class="inline-block mr-2" />
        <span>{{ data.copyright }}</span>
      </UiP>
    </div>
  </div>
</template>
