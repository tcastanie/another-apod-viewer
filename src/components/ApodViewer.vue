<template>
  <LoadingLayer v-if="loading" />
  <ErrorLayer v-else-if="errorMessage">{{ errorMessage }}</ErrorLayer>
  <div v-else class="c-apod-viewer">
    <PictureViewer
      v-if="apod.media_type === 'image'"
      :img-url="apod.url"
      :full-url="apod.hdurl"
    />
    <VideoViewer
      v-else-if="apod.media_type === 'video'"
      :video-url="apod.url"
    />
    <div class="headings">
      <h1>{{ apod.title }}</h1>
      <h2>{{ apod.date }}</h2>
    </div>
    <div class="c-apod-viewer__description"></div>
    <div class="c-apod-viewer__copyright"></div>
    <blockquote>
      <b>Explanation:</b> {{ apod.explanation }}
      <footer v-if="apod.copyright">
        <cite>Â© {{ apod.copyright }}</cite>
      </footer>
    </blockquote>
  </div>
</template>

<script>
import { useStore } from "@/stores/apod";
import LoadingLayer from "./LoadingLayer.vue";
import ErrorLayer from "./ErrorLayer.vue";
import PictureViewer from "./PictureViewer.vue";
import VideoViewer from "./VideoViewer.vue";
export default {
  components: {
    LoadingLayer,
    ErrorLayer,
    PictureViewer,
    VideoViewer,
  },
  props: {
    dateParam: {
      type: String,
      default: null,
    },
  },
  setup() {
    const store = useStore();
    return { store };
  },
  data() {
    return {
      loading: false,
      errorMessage: null,
      apod: {},
    };
  },
  created() {
    if (this.store.apodsAlreadyFetched[this.dateParam]) {
      this.apod = this.store.apodsAlreadyFetched[this.dateParam];
    } else {
      this.fetchApod();
    }
  },
  methods: {
    fetchApod() {
      this.loading = true;
      const url = this.dateParam
        ? this.store.getUrlWithDate(this.dateParam)
        : this.store.getTodayUrl;
      this.fetchUrl(url);
    },
    fetchUrl(url) {
      fetch(url)
        .then(async (response) => {
          const data = await response.json();
          if (!response.ok) {
            const error1 = data && (data.msg || data.error.code);
            const error2 = `${response.statusText} (${response.status})`;
            return Promise.reject(new Error(`${error1}. ${error2}`));
          }
          this.apod = data;
          this.loading = false;
        })
        .catch((error) => {
          this.errorMessage = error;
          this.loading = false;
        });
    },
  },
};
</script>
