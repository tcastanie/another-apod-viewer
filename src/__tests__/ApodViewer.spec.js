import { describe, it, expect } from "vitest";
import { mount } from "@vue/test-utils";
import ApodViewer from "@/components/ApodViewer.vue";
import { createTestingPinia } from "@pinia/testing";
import { testList } from "./constantsMockups";

describe("ApodViewer", () => {
  it("renders properly", () => {
    const wrapper = mount(ApodViewer, {
      global: {
        plugins: [createTestingPinia()],
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders properly", () => {
    const wrapper = mount(ApodViewer, {
      global: {
        plugins: [
          createTestingPinia({
            initialState: {
              apod: { apodsAlreadyFetched: testList[0] },
            },
          }),
        ],
      },
    });
    // TODO
    expect(wrapper.html()).toMatchSnapshot();
  });
});
