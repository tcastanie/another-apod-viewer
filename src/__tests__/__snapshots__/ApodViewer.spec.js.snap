// Vitest Snapshot v1

exports[`ApodViewer > renders properly 1`] = `"<article aria-busy=\\"true\\" class=\\"c-loading\\" data-v-4320971b=\\"\\">Loading...</article>"`;

exports[`ApodViewer > renders properly 2`] = `"<article aria-busy=\\"true\\" class=\\"c-loading\\" data-v-4320971b=\\"\\">Loading...</article>"`;
